File: prng.rb
PRNG = Struct.new(:seed) # ==, eql?, hashのためにStructから作る

PRNG.class_eval do
  def rand(n)
    p = dup()
    [p, p.rand!(n)]
  end

  def rand!(n)
    succ!
    (seed >> 16) % n
  end
  
  def succ!
    self.seed = (seed * 0x41c64e6d + 0x6073) & 0xffffffff
  end
end

File: js-version/prng.js
class PRNG {
	constructor(seed) {
		this.seed = seed;
	}
	rand(n) {
		let prngp = this.dup();
		return [prngp, prngp.randQ(n)];
	}
	randQ(n) {
		this.succ();
		return (this.seed >>> 16) % n;
	}
	succ() {
		const A = 0x41c64e6d, B = 0x6073;
		this.seed = (this._mul(this.seed, A) + B) >>> 0;
	}
	dup() {
		return new PRNG(this.seed);
	}
	_mul(a, b) {
		let a1 = a >>> 16, a2 = a & 0xffff;
		let b1 = b >>> 16, b2 = b & 0xffff;
		return (((a1 * b2 + a2 * b1) << 16) + a2 * b2) >>> 0;
	}
}

